### PR TITLE
[IA-1826] Kubernetes app resource

### PR DESCRIPTION
**Pull request by akarukappadath**
Ticket: https://broadworkbench.atlassian.net/browse/IA-1826
Design doc: https://broadworkbench.atlassian.net/wiki/spaces/IA/pages/595492869/2020-07-07+Kubernetes+App+Sam+Model
<Put notes here to help reviewer understand this PR>

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

**Comment by akarukappadath** [comment]
jenkins retest

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -356,6 +365,32 @@ resourceTypes = {
     }
     reuseIds = false
   }
+  kubernetes-app = {
```
Comment (rtitle) [inline src/main/resources/reference.conf:356]: Commented in the design doc as well, but I'm thinking we can:
* drop the `user` role -- there's no distinction between app creator and user; only the creator can use the app (similar to notebook runtimes).
* remove the `stop` and `start` actions -- Leo won't expose this functionality for k8s
* add a `connect` action -- only the app creator should be able to connect to the app, whereas the project owner can delete but not connect.

Let me know if I'm misunderstanding something.

File: src/main/resources/reference.conf [hunk 2]
```diff
@@ -306,9 +312,6 @@ resourceTypes = {
       connect = {
         description = "connect to the Jupyter notebook running on the notebook cluster"
       }
-      sync = {
```
Comment (rtitle) [inline src/main/resources/reference.conf:309]: Context: we're not using this action; `connect` covers data syncing as well

**Review by rtitle — APPROVED** [review]
This LGTM as we've discussed -- Cloud Integration team should review as well though.

File: src/main/resources/reference.conf [hunk 3]
```diff
@@ -262,6 +259,15 @@ resourceTypes = {
...
         description = "delete persistent disk in this billing-project"
       }
+      list_kubernetes_app = {
+        description = "list all kubernetes applications in this billing-project"
+      }
+      create_kubernetes_app = {
+        description = "create kubernetes application in this billing-project"
+      }
+      delete_kubernetes_app = {
+        description = "delete kubernetes application in this billing-project"
```
Comment (dvoet) [inline src/main/resources/reference.conf]: this is delete _any_ k8s app in the billing-project, correct? But the billing-project owner won't have delete access on the kubernetes-app resource so they won't be able to remove it from sam.

File: src/main/resources/reference.conf (hunk 1 above)
Comment (dvoet) [reply src/main/resources/reference.conf:356]: where is the design doc?

Comment (akarukappadath) [reply src/main/resources/reference.conf:356]: https://broadworkbench.atlassian.net/wiki/spaces/IA/pages/595492869/2020-07-07+Kubernetes+App+Sam+Model

**Review by dvoet — COMMENTED** [review]
I feel like the only `billing-project` level action should be `create_kubernetes_app`. The `kubernetes_app` resource type should have another role, `manager`, which would be able to delete the app. A list action is not required, just list all the `kubernetes_app`s a user has access to. At the policy level, `kubernetes_app` resources should be created with a creator policy _and_ a manager policy whose member is the email address of the billing-project owner policy. This way we do not infer access to one resource based on access to another.

**Comment by rtitle** [comment]
Interesting. That makes sense as a way to model this. Our other Leo resources (like runtimes, disks) also follow the pattern in this PR. I wonder if we should consider migrating those too (probably separately from the k8s work).

**Comment by akarukappadath** [comment]
jenkins retest

**Review by dvoet — APPROVED** [review]

**Comment by rtitle** [comment]
@dvoet, question about this:
> At the policy level, kubernetes_app resources should be created with a creator policy and a manager policy whose member is the email address of the billing-project owner policy.

It just occurred to me, if the app is being created by a user who is NOT the billing-project owner, how can we get the billing-project owner policy email? The calling user does not have permission to look it up in Sam. Can the Leo SA be used to look up arbitrary billing-project policies?

**Pull request merged**